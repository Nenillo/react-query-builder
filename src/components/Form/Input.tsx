import React from 'react';
import styled from 'styled-components';
import { colors } from '../../constants/colors';

const StyledInput = styled.input`
  padding: 0.2rem 0.5rem;
  height: 1.6rem;
  line-height: 1.6rem;
  border: 1px solid ${colors.medium};
  border-radius: 3px;
`;

export interface InputProps {
  type: 'date' | 'number' | 'text';
  value: string;
  onChange: (value: string) => void;
  className?: string;
}

export const Input: React.FC<InputProps> = ({
  type,
  value,
  onChange,
  className,
}) => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value);
  };

  return (
    <StyledInput
      type={type}
      value={value}
      onChange={handleChange}
      className={className}
    />
  );
};
