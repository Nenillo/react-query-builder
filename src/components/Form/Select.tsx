import React from 'react';

import styled from 'styled-components';
import { colors } from '../../constants/colors';

const StyledSelect = styled.select`
  padding: 0.2rem 0.5rem;
  height: 1.6rem;
  line-height: 1.6rem;
  border: 1px solid ${colors.medium};
  border-radius: 3px;
`;

export interface SelectProps {
  values: { value: React.ReactText; label: string }[];
  selectedValue?: React.ReactText;
  emptyValue?: string;
  onChange: (value: any) => void;
  className?: string;
}

export const Select: React.FC<SelectProps> = ({
  values,
  selectedValue,
  emptyValue,
  onChange,
  className,
}) => {
  const handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    onChange(event.target.value);
  };

  return (
    <StyledSelect
      onChange={handleChange}
      value={selectedValue}
      className={className}
    >
      <option value="" disabled>
        {emptyValue}
      </option>
      {values.map(({ value: optionValue, label: optionLabel }) => (
        <option value={optionValue} key={optionValue}>
          {optionLabel}
        </option>
      ))}
    </StyledSelect>
  );
};
