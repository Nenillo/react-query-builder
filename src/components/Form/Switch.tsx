import React from 'react';
import styled from 'styled-components';
import { colors } from '../../constants/colors';

const Knob = styled.div`
  position: absolute;
  width: 1.3rem;
  height: 1.3rem;
  background: white;
  border-radius: 50%;
  border: 1px solid ${colors.dark};
  box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.4);
`;

const StyledSwitch = styled.div<{ switched: boolean }>`
  width: 3rem;
  height: 1.65rem;
  position: relative;
  border-radius: 1.4rem;
  border: 1px solid ${colors.dark};
  background-color: ${({ switched }) =>
    switched ? colors.primary : colors.darker};
  transition: all 0.5s;
  cursor: pointer;

  ${Knob} {
    transition: all 0.5s;
    top: 0.1rem;
    left: ${({ switched }) => (switched ? '1.3rem' : '0.1rem')};
  }
`;

export interface SwitchProps {
  switched: boolean;
  onChange?: (value: boolean) => void;
  className?: string;
}

export const Switch: React.FC<SwitchProps> = ({
  switched,
  onChange,
  className,
}) => {
  const handleClick = () => {
    if (onChange) {
      onChange(!switched);
    }
  };

  return (
    <StyledSwitch
      data-test="Switch"
      switched={switched}
      onClick={handleClick}
      className={className}
    >
      <Knob />
    </StyledSwitch>
  );
};
