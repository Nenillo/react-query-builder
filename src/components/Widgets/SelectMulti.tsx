import React, { useContext } from 'react';
import { BuilderContext } from '../Builder';
import { clone } from '../../utils/clone';

export interface SelectMultiProps {
  values: { value: string; label: string }[];
  selectedValue: string[];
  id: string;
}

export const SelectMulti: React.FC<SelectMultiProps> = ({
  values,
  selectedValue,
  id,
}) => {
  const { data, setData, onChange, components, strings } = useContext(
    BuilderContext
  );

  const { form } = components;

  const handleChange = (value: string) => {
    if (setData && onChange) {
      const clonedData = clone(data);
      const parentIndex = clonedData.findIndex((item: any) => item.id === id);

      clonedData[parentIndex].value = clonedData[parentIndex].value.filter(
        (item: any) => item !== value
      );
      clonedData[parentIndex].value.push(value);

      setData(clonedData);
      onChange(clonedData);
    }
  };

  const handleDelete = (value: string) => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);

    clonedData[parentIndex].value = clonedData[parentIndex].value.filter(
      (item: any) => item !== value
    );

    setData(clonedData);
    onChange(clonedData);
  };

  if (form && strings.form) {
    return (
      <form.SelectMulti
        onChange={handleChange}
        onDelete={handleDelete}
        selectedValue={selectedValue}
        emptyValue={strings.form.selectYourValue}
        values={values}
      />
    );
  }

  return null;
};
