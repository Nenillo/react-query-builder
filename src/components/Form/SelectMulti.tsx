import React from 'react';
import styled from 'styled-components';
import { colors } from '../../constants/colors';
import { Select, SelectProps } from './Select';

const Container = styled.div`
  display: grid;
  grid-auto-columns: min-content;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
`;

const OptionContainer = styled.div`
  display: grid;
  grid-auto-columns: min-content;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
  align-self: center;
`;

const Option = styled.span`
  padding: 0.3rem 0.5rem;
  color: ${colors.dark};
  font-size: 0.7rem;
  line-height: 0.7rem;
  white-space: nowrap;
  border: 1px solid ${colors.dark};
  border-radius: 3rem;
`;

const Delete = styled.span`
  cursor: pointer;
`;

export interface SelectMultiProps
  extends Pick<SelectProps, 'onChange' | 'values'> {
  onDelete: (value: string) => void;
  selectedValue: string[];
  emptyValue?: string;
}

export const SelectMulti: React.FC<SelectMultiProps> = ({
  onChange,
  onDelete,
  selectedValue,
  emptyValue,
  values,
}) => {
  return (
    <Container>
      <Select
        onChange={onChange}
        selectedValue=""
        emptyValue={emptyValue}
        values={values}
      />
      <OptionContainer>
        {selectedValue.map((value, key) => {
          const labelIndex = values.findIndex(item => item.value === value);

          return (
            <Option key={key}>
              {values[labelIndex].label}{' '}
              <Delete onClick={() => onDelete(value)} data-test="Delete">
                [x]
              </Delete>
            </Option>
          );
        })}
      </OptionContainer>
    </Container>
  );
};
