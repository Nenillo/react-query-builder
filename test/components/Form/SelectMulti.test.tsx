import { SelectMulti } from '../../../src/components/Form/SelectMulti';
import { mount, shallow } from 'enzyme';
import React from 'react';

const mockValues = [{ value: 'test', label: 'test' }];

describe('#components/SelectMulti', () => {
  it('Tests Snapshot', () => {
    expect(
      shallow(
        <SelectMulti
          disabled={false}
          onChange={jest.fn()}
          onDelete={jest.fn()}
          selectedValue={['test']}
          values={mockValues}
        />
      )
    ).toMatchSnapshot();
  });

  it('Tests user interaction', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <SelectMulti
        disabled={false}
        onChange={onChange}
        onDelete={jest.fn()}
        selectedValue={['test']}
        values={mockValues}
      />
    );

    const select = wrapper.find('select');

    select.simulate('focus');
    select.simulate('change', { target: { value: ['test'] } });
    select.simulate('change', { target: { value: [] } });
    expect(onChange).toBeCalledTimes(2);
  });
});
