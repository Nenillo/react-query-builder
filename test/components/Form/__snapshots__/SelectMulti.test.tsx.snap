// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#components/SelectMulti Tests Snapshot 1`] = `
<styled.div>
  <Component
    disabled={false}
    onChange={[MockFunction]}
    selectedValue=""
    values={
      Array [
        Object {
          "label": "test",
          "value": "test",
        },
      ]
    }
  />
  <styled.div>
    <styled.span
      key="0"
    >
      test
       
      <styled.span
        data-test="Delete"
        disabled={false}
        onClick={[Function]}
      >
        [x]
      </styled.span>
    </styled.span>
  </styled.div>
</styled.div>
`;
